xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -651,163 +1258,177 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
+    for (unsigned int j = run.startOffset; j < run.endOffset; j++)
     {
-      hb_buffer_add(run.m_buffer, static_cast<wchar_t>(0xffff & text[j]), j);
+      hb_buffer_add(buffer, text[j].letter, j);
     }
 
-    hb_buffer_set_content_type(run.m_buffer, HB_BUFFER_CONTENT_TYPE_UNICODE);
-    hb_shape(m_hbFont, run.m_buffer, nullptr, 0);
+    if (run.font)
+      hb_shape(run.font->m_hbFont, buffer, nullptr, 0);
+
     unsigned int glyphCount;
-    run.m_glyphInfos = hb_buffer_get_glyph_infos(run.m_buffer, &glyphCount);
-    run.m_glyphPositions = hb_buffer_get_glyph_positions(run.m_buffer, &glyphCount);
+    run.glyphInfos = hb_buffer_get_glyph_infos(buffer, &glyphCount);
+    run.glyphPositions = hb_buffer_get_glyph_positions(buffer, &glyphCount);
 
     for (size_t k = 0; k < glyphCount; k++)
     {
-      glyphs.emplace_back(run.m_glyphInfos[k], run.m_glyphPositions[k]);
+      glyphs.emplace_back(run.glyphInfos[k], run.glyphPositions[k], run.font);
     }
-
-    hb_buffer_destroy(run.m_buffer);
   }
+  hb_buffer_destroy(buffer);
 
   return glyphs;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr, FT_UInt glyphIndex)
+Character* CGUIFontTTF::GetCharacter(const character_t& chr,
+                                     FT_UInt glyphIndex,
+                                     CFontData* fontData)
 {
-  const wchar_t letter = static_cast<wchar_t>(chr & 0xffff);
+  typedef uint32_t char_quik_t;
+
+  const auto letter = chr.letter;
 
   // ignore linebreaks
-  if (letter == L'\r')
+  if (letter == static_cast<decltype(chr.letter)>('\r'))
     return nullptr;
 
-  const character_t style = (chr & 0x7000000) >> 24; // style = 0 - 6
+  const auto style = chr.style;
+
+  if (!fontData)
+  {
+    glyphIndex = FT_Get_Char_Index(m_fontMain.m_face, chr.letter);
+    if (glyphIndex != 0)
+      fontData = &m_fontMain;
+    else
+    {
+      const CFontTable* fontTable;
+      fontData = FindFallback(chr.letter, fontTable);
+      if (!fontData || (glyphIndex = FT_Get_Char_Index(fontData->m_face, chr.letter)) != 0)
+      {
+        CLog::Log(LOGERROR,
+                  "CGUIFontTTF::{}: Unable to find font about character {:x}, font data found: {}",
+                  __func__, static_cast<uint32_t>(chr.letter), fontData ? "yes" : "no");
+        return nullptr;
+      }
+    }
+  }
 
   if (!glyphIndex)
-    glyphIndex = FT_Get_Char_Index(m_face, letter);
+    glyphIndex = FT_Get_Char_Index(fontData->m_face, chr.letter);
 
   // quick access to the most frequently used glyphs
-  if (glyphIndex < MAX_GLYPH_IDX)
+  if (glyphIndex < MAX_GLYPH_IDX && fontData->m_mainFont)
   {
-    character_t ch = (style << 12) | glyphIndex; // 2^12 = 4096
+    char_quik_t ch = (style << 12) | glyphIndex; // 2^12 = 4096
 
     if (ch < LOOKUPTABLE_SIZE && m_charquick[ch])
       return m_charquick[ch];
   }
 
   // letters are stored based on style and glyph
-  character_t ch = (style << 16) | glyphIndex;
+  const glyph_and_style_t ch =
```

Comment:
I do not understand much this change: `style` is in range 0 to 6 and `glyphIndex` 0 to 65536 because TT / OpenType fonts can only store 65536 glyphs (AFAIK). Then to map this in one variable should be:

```
style           ch (depending of glyphIndex)
 0 --------->       0 to 65,536
 1 --------->  65,637 to 131,072
 2 ---------> 131,073 to 196,609
 3
 4
 5
 6 ---------> 393,216 to 458,752
```
And the formula to accomplish this is:

ch = (style * 65536) + glyphIndex

If we want use bits shift method (because is supposed is more optimized code):

`const uint32_t ch = (style << 16) | glyphIndex;`       

(2^16 = 65536)

And this is what was there before : D

I also think it's debatable whether it's necessary to use bit shifting instead of multiplication since it's supposed to be the compiler's job to optimize the code for each platform and there's no use wanting to do it manually because the generated assembly code will probably be the same.

It is even possible that the compiler will do better if we express what we want to do and not how to do it.

Also is not need use `int64_t` for `glyph_and_style_t` because max range is 458,752 or maybe I'm not taking something new into account?

Reply:
Just as an FYI, there is a spec proposal to remove the 65k glyph limit. https://github.com/harfbuzz/boring-expansion-spec

Reply:
Yes, that's seen during search on web. A 16bit 0xFFFF can no more enough for all cases.

Reply:
About the shift, yes is better to use an "<< 32", thought first to define the size as this to have more clean on changes, but makes no sense ☺️.